### PR TITLE
[build] Update to .NET 6 Preview 4

### PR DESCRIPTION
**Pull request by pjcollins**
Brings in Preview 4 updates for the Android, MaciOS, and .NET SDks.

Initial build attempts were failing with hundreds of errors:

    D:\a\1\s\src\Essentials\src\Connectivity\Connectivity.ios.tvos.macos.cs(12,70): error BI1234: 'CTCellularData' is obsolete: 'Starting with ios14.0 Use the 'CallKit' API instead.' [D:\a\1\s\src\Essentials\src\Essentials-net6.csproj]
    D:\a\1\s\src\Essentials\src\Compass\Compass.ios.cs(31,6): error CA1416: This call site is reachable on: 'iOS' 13.6 and later. 'CLLocationManager.HeadingFilter' is only supported on: 'macos' 11.0 and later. [D:\a\1\s\src\Essentials\src\Essentials-net6.csproj]
    D:\a\1\s\src\Core\src\Platform\MaciOS\ColorExtensions.cs(64,13): error CA1416: This call site is reachable on: 'MacCatalyst' 13.5 and later. 'UIColor.SystemBackgroundColor.get' is supported on: 'ios' 13.0 and later. [D:\a\1\s\src\Core\src\Core-net6.csproj]
    CSC : error AD0001: Analyzer 'Microsoft.NetCore.Analyzers.InteropServices.PlatformCompatibilityAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'. [D:\a\1\s\src\BlazorWebView\src\core\Microsoft.AspNetCore.Components.WebView.Maui.csproj]

The [platform compatibility analyzer][0] has been disabled for now, and
it can be enabled locally to work through the few hundred errors in
future PRs.  The `BI1234` instances have been converted to warnings.

Once these issues were sorted, there were a handful of other breaks
that I've attempted to fix:

    D:\a\1\s\src\Essentials\src\AssemblyInfo\AssemblyInfo.ios.tvos.watchos.macos.cs(3,12): error CS0618: 'LinkerSafeAttribute' is obsolete: 'Replace with '[assembly: System.Reflection.AssemblyMetadata ("IsTrimmable", "True")]'.' [D:\a\1\s\src\Essentials\src\Essentials-net6.csproj]

    D:\a\1\s\src\Compatibility\Core\src\iOS\Renderers\TabbedRenderer.cs(422,33): error CS1503: Argument 1: cannot convert from 'UIKit.UIStringAttributes' to 'UIKit.UITextAttributes' [D:\a\1\s\src\Compatibility\Core\src\Compatibility-net6.csproj]

    D:\a\1\s\src\Compatibility\Core\src\iOS\CollectionView\CarouselViewController.cs(106,23): error CS0114: 'CarouselViewController.DraggingStarted(UIScrollView)' hides inherited member 'UICollectionViewController.DraggingStarted(UIScrollView)'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [ D:\a\1\s\src\Compatibility\Core\src\Compatibility-net6.csproj]
    D:\a\1\s\src\Compatibility\Core\src\iOS\CollectionView\CarouselViewController.cs(117,23): error CS0114: 'CarouselViewController.DraggingEnded(UIScrollView, bool)' hides inherited member 'UICollectionViewController.DraggingEnded(UIScrollView, bool)'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [ D:\a\1\s\src\Compatibility\Core\src\Compatibility-net6.csproj]

The linker has been disabled for MaciOS projects until we are able
to upgrade to Big Sur and Xcode 12.5.

Finally it seems JDK 14 was being preferred on macOS machines in CI, so
I've set `$(JI_JAVA_HOME)` to the [pre-installed JDK 11][1].

[0]: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1416
[1]: https://github.com/actions/virtual-environments/blob/macOS-10.15/20210419.2/images/macos/macos-10.15-Readme.md#java

### CODE REVIEW

**Comment by pjcollins** [comment]
How should we handle the Xcode 12.5 beta bump here? Do we want to provision this or should we disable the linker as needed?

**Comment by jonathanpeppers** [comment]
I would say we should install Xcode 12.5 beta in the `build_net6` stage:
```
error MT2301: The linker step 'Setup' failed during processing: This version of Microsoft.iOS requires the iOS 14.5 SDK (shipped with Xcode 12.5).
```

There is a `xcode.csx` that appears to use this `$REQUIRED_XCODE` env var?

File: src/Compatibility/Core/src/iOS/CollectionView/CarouselViewController.cs [hunk 1]
```diff
@@ -103,21 +103,17 @@ void BoundsSizeChanged()
 
 #if MACCATALYST
 		[Foundation.Export("scrollViewWillBeginDragging:")]
```
Comment (rolfbjarne) [inline src/Compatibility/Core/src/iOS/CollectionView/CarouselViewController.cs:105]: You should be able to remove this Export attribute as well now.

File: src/Compatibility/Core/src/iOS/CollectionView/CarouselViewController.cs [hunk 2]
```diff
@@ -103,21 +103,17 @@ void BoundsSizeChanged()
...
-		public override void DraggingStarted(UIScrollView scrollView)
 #endif
+		public override void DraggingStarted(UIScrollView scrollView)
 		{
 			_isDragging = true;
 			Carousel.SetIsDragging(true);
 		}
 
 #if MACCATALYST
 		[Foundation.Export("scrollViewDidEndDragging:willDecelerate:")]
```
Comment (rolfbjarne) [inline src/Compatibility/Core/src/iOS/CollectionView/CarouselViewController.cs:114]: Same: this can probably be removed.

**Review by rolfbjarne — APPROVED** [review]

**Comment by pjcollins** [comment]
Unfortunately, trying to use the [`suggestion` severity level][0] for platform compatibility violations didn't seem to work for me locally. It also resulted in these NREs returning, which would need to be converted to warnings or suppressed separately:

>   CSC : error AD0001: Analyzer 'Microsoft.NetCore.Analyzers.InteropServices.PlatformCompatibilityAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'. [C:\Users\Peter\source\pj\maui\src\Essentials\src\Essentials-net6.csproj]

I think the single entry in `.editorconfig` that can be modified locally to fix things may be the best path for now, unless we want every build producing hundreds of warnings. 

[0]: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-options#severity-level

File: GitInfo.txt [hunk 3]
```diff
@@ -1 +1 @@
-6.0.100-preview.3
+6.0.100-preview.4
```
Comment (pjcollins) [inline GitInfo.txt:1]: Is there something else I should do here to reset the commit distance for the version number?

**Comment by pjcollins** [comment]
Looks like Xcode 12.5 requires Big Sur, and Microsoft hosted images for Big Sur [won't be available until the summer][0].

Do we know if there are any internal Big Sur pools that we may be able to migrate to? @akoeplinger mentioned that there may be some Helix options but I am not sure if we can build on those or what that migration would look like.

We can try to get a pool set up if one doesn't exist but I imagine this will take time.  Should we disable linking for iOS for now?

[0]: https://github.com/actions/virtual-environments/issues/2072#issuecomment-823881568

**Comment by jonathanpeppers** [comment]
@pjcollins in this file:

https://github.com/dotnet/maui/blob/a00c135acc6e2cc7ef41c964e0e5d2f83e72d4a5/Directory.Build.targets#L1-L6

Try something like:

```xml
<!-- Temporarily disable the linker for net6.0-ios until we get Xcode 12.5 and Big Sur on the Hosted Pool -->
<PropertyGroup>
  <LinkMode Condition=" '$(TargetFramework)' == 'net6.0-ios' ">None</LinkMode>
</PropertyGroup>
```

**Comment by pjcollins** [comment]
This may have been a false alarm actually, I think we may have a non hosted option we can use here.

File: GitInfo.txt (hunk 3 above)
Comment (rmarinho) [reply GitInfo.txt:1]: Nop , we use that as  the final version number

**Review by jonathanpeppers — APPROVED** [review]

**Review by Redth — APPROVED** [review]

**Comment by Eilon** [comment]
FYI: If anyone sees anything broken with BlazorWebView, please let me know. There's a breaking change in .NET 6 Preview 4 (from Preview 3) but I don't _think_ that this PR would pull that in. I'll pull in that dependency later.

File: GitInfo.txt (hunk 3 above)
Comment (rmarinho) [reply GitInfo.txt:1]: i think we would this on the release branch no?

**Review by rmarinho — APPROVED** [review]

**Pull request merged**